### DISCUSSION TITLE
MFEM for compressible real-gas LES?

### DISCUSSION THREAD

**Discussion by okcfdlab** [mfem/mfem#2707] [Q&A]
Hello MFEM community, I am investigating the blending of hydrogen into natural gas pipelines. One of the important questions is of course safety associated with the hydrogen gas. High-pressure releases of hydrogen into an oxygen-rich environment has the potential to result in an explosive detonation of the hydrogen. This can occur when the release occurs through an orifice due to a puncture of a gas system component. The released hydrogen gas produces a shock wave in the air driven by the piston action of the expanding hydrogen jet, compressing and heating the air. The diffused interface between the hot air and hydrogen may lead to a local spontaneous ignition (auto-ignition) without any external sources. To date, there has been no research into the auto-ignition risk of hydrogen-enriched natural gas. Although the storage pressures in natural gas lines are typically below the pressures required for pure hydrogen auto-ignition (close to 200 atm), sufficiently large storage pressures will be expected if the hydrogen-enriched natural gas is used in compressed natural gas (CNG) cylinders e.g. in automotive applications, where pressures exceeding 300 atm may be expected.

In addition to detonation risk, understanding the flammability zones of the gas following an accidental release of hydrogen-enriched natural gas is critical for ensuring safety. While natural gas has a relatively narrow flammability range in air, hydrogen has a much broader range, meaning that current flame safety zones may need to be altered in response to an accidental release. To evaluate this risk, we require a better understanding of how a blended mixture of hydrogen and natural gas will disperse following accidental release. Prior studies have shown that the shape of the puncture orifice, the jet orientation, and buoyancy effects can significantly alter the trajectories of subsonic and supersonic hydrogen jets. Real-gas effects may also influence the flammability region, such as Joule-Thomson cooling of the methane during expansion, which may alter the buoyancy forces and affect the gas dispersion. As no studies to date have investigated these effects in the context of mixed hydrogen-methane jets, the impact of hydrogen-enrichment on the flammability safety envelope following accidental release is unclear.

I'm investigating the suitability of MFEM for developing a bespoke CFD tool for conducting simulations of this problem. I think the following features will be important:

- High-order discretization to ensure numerical accuracy and scalability
- Accurate resolution of shock waves
- Incorporation of real-gas effects through a non-ideal EOS
- A miscible mixture of independent gas species
- At present, I will ignore combustion reactions, though that will needed in the future
- Turbulent dispersion will be important but the Reynolds numbers will be large, thus necessitating large eddy simulation (LES) approach

My questions:

1. Is MFEM suitable for such a problem?
2. Have others used MFEM for developing real-gas compressible LES solvers?
3. Where does the community think I should begin in terms of learning? I notice that Laghos miniapp has some of the desired features--is that a reasonable starting point?

Thank you!

**Comment by fab6** [comment]
Hi,

did you find a solution yet? Actually I am wondering as well... Maybe Blast is something similar, but I am not sure, if this is available for the public.

Best 
Fab

**Reply by okcfdlab** [reply]
Hi Fab, I didn't get a full solution yet. I'm slowly working my way through learning MFEM, trying to see how suitable it is for my problem. If you have any insight, it would be appreciated.

I believe that BLAST is not available publicly, or at least not to non-USA researchers--I'm in Canada--although I may be mistaken.

Cheers,

Joshua

**Reply by tzanio** [reply]
You are correct, BLAST is not open source

**Comment by fab6** [comment]
Ok, interesting, yes, this is what I thought about Blast; not sure how active the community is outside Blast for something  similar as open-source.
But I am extremely far from being an expert here and I am just curious of potential high order exa-scale libraries for general low speed fire simulation. Maybe exadg/deal.ii or pyfr are alternatives for you, but they would need maybe similar effort to get what your are looking for...not sure.

Best 
Fab

**Reply by okcfdlab** [reply]
Thanks. Yes, I've looked at deal.ii and PyFR, along with these:
https://www.ukctrf.com/index.php/senga/
https://www.ukctrf.com/index.php/hamish/
https://www.incompact3d.com/about.html
https://www.dealii.org/
https://mfem.org/
https://su2code.github.io/tutorials/NICFD_nozzle/
https://numath.dmae.upm.es/tools/horses3d/
https://www.nektar.info/

Not all have the required flexibility, which is why I'm still thinking about which platform to build the required functionality.

**Comment by fab6** [comment]
ah, cool,  maybe just to add just for information:
https://github.com/exadg/exadg
https://amrex-combustion.github.io/

How important is GPU for you?

**Reply by okcfdlab** [reply]
GPU would be nice but not essential

**Comment by tzanio** [comment]
@fab6 and @okcfdlab,

MFEM is a mathematical library, not a physics code.

Questions like this are out-of-scope here.

Tzanio

**Reply by fab6** [reply]
yes, that's right, we drifted a bit away of the initial topic; excuse me for this.

**Comment by v-dobrev** [comment]
Hi @okcfdlab,

To answer your original questions:

> My questions:
>     1. Is MFEM suitable for such a problem?
>     2. Have others used MFEM for developing real-gas compressible LES solvers?
>     3. Where does the community think I should begin in terms of learning? I notice that Laghos miniapp has some of the desired features--is that a reasonable starting point?

1. Depending on the discretization approach you want to use, MFEM could be suitable. For example, if you want to use unstructured meshes and finite element methods, MFEM is suitable, in general. If you want to focus on structured meshes with finite differences or finite volumes, other libraries may be a better choice.
2. For the type of problems that you are describing, I do not know of any open source projects based on MFEM.
3. This also depends to a large extend on the type of discretization you want to use. One possible example of interest is example 18/18p, see https://mfem.org/examples.
